glib/src/gstring.rs

```diff
@@ -660,23 +660,23 @@ impl From<GString> for Box<str> {
     }
 }
 
-impl From<String> for GString {
+impl TryFrom<String> for GString {
```

Comment:
@bilelmoussaoui What do you think?

Reply:
Any opinions? :)

Reply:
I guess it is fine to also keep a From implementation, but those are not expected to panic, aren't they?

Reply:
Yes but then you're also not expecting `label.set_text("bla\0bla")` to panic :)

Reply:
@jf2048 Any opinions from your side here?

Reply:
Is there any harm in passing through a string with nul bytes? That would be faster if we could just skip the check altogether. Sure the C API will truncate it, but is that a worse behavior than panicking?

Reply:
> but is that a worse behavior than panicking?

Depends on what the string is used for, I guess :)

> Otherwise I don't really see a problem with implementing a panicking `From`

Let's go with that then for now?

Reply:
Let's wait a bit on this, I want to test out this one first: https://github.com/gtk-rs/gir/pull/1184

Reply:
My idea for the ideal approach:
- Implement both `From` and `TryFrom` so users can get the fallible constructor, if they want.
- In debug builds, the `From` impl will scan the string during construction and panic on interior nuls. The check will be skipped in production builds. Also do this in `impl ToGlibPtr for str`, etc.
- Since the `CString` invariant will no longer hold in all cases, change `gstring::Inner::Native` to just contain a `String`. Also remove `AsRef<CStr>` and add a `to_cstr` method that does the same thing as [`CStr::from_bytes_until_nul`](https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.from_bytes_until_nul).

Reply:
Sounds good to me

Reply:
This would also solve https://github.com/gtk-rs/gtk-rs-core/issues/10

Reply:
A couple more notes.

- Since it's immutable, it should be `Native(Option<Box<str>>)` to preserve the same memory layout.
- We can support `Native(None)` as a valid case to optimize an empty string, that way it can skip an allocation when calling `to_glib_none` on an empty `GString`.

Reply:
`NULL` and `""` are different things though, and passing the former instead of the latter can potentially cause some C code to crash

Reply:
It won't return `NULL`, it will always produce a static `""`.

Reply:
Strings now using `From` again,.`TryFrom` is only used for conversion that could have invalid UTF-8.